source/common/buffer/zero_copy_input_stream_impl.cc

```diff
@@ -0,0 +1,54 @@
+#include "common/buffer/zero_copy_input_stream_impl.h"
+
+#include "common/common/assert.h"
+
+namespace Envoy {
+namespace Buffer {
+
+ZeroCopyInputStreamImpl::ZeroCopyInputStreamImpl(Buffer::Instance& buffer) {
+  move(buffer);
+  finish();
+}
+
+void ZeroCopyInputStreamImpl::move(Buffer::Instance& instance) {
+  ASSERT(!finished_);
+
+  buffer_.move(instance);
+}
+
+bool ZeroCopyInputStreamImpl::Next(const void** data, int* size) {
+  if (position_ != 0) {
+    buffer_.drain(position_);
+    position_ = 0;
+  }
+
+  Buffer::RawSlice slice;
+  const uint64_t num_slices = buffer_.getRawSlices(&slice, 1);
+
+  if (num_slices > 0) {
+    *data = slice.mem_;
+    *size = slice.len_;
+    position_ = slice.len_;
+    byte_count_ += slice.len_;
+    return true;
+  }
+
+  if (!finished_) {
```

Comment:
I'm a little confused what happens here. If we return true here with no data, how does the proto code know to call Next() again without just going into a spin loop? Or is this only useful in the explicit transcoding case somehow? Again might need a few more comments for those not as familiar with proto interfaces.